docs/proposals/autoscaling.md

```diff
@@ -0,0 +1,245 @@
+with scaling based on traffic, predictive analysis may be used to determine anticipated system load and scale the application automatically.  
+
+### Scaling based on arbitrary data
+
+Administrators may wish to scale the application based on any number of arbitrary data points such as job execution time or
+duration of active sessions.  There are any number of reasons an administrator may wish to increase or decrease capacity which 
+means the auto-scaler must be a configurable, extensible component.
+
+## Specification
+
+In order to facilitate talking about auto-scaling the following definitions are used:
+
+* `ReplicationController` - the first building block of auto scaling.  Pods are deployed and scaled by a `ReplicationController`.
+* kube proxy - The proxy handles internal inter-pod traffic, an example of a data source to drive an auto-scaler
+* L3/L7 proxies - A routing layer handling outside to inside traffic requests, an example of a data source to drive an auto-scaler
+* auto-scaler - scales replicas up and down by using the `resize` endpoint provided by scalable resources (`ReplicationController`)
+
+
+### Auto-Scaler
+
+The Auto-Scaler is a state reconciler responsible for checking data against configured scaling thresholds 
+and calling the `resize` endpoint to change the number of replicas.  The scaler will 
+use a client/cache implementation to receive watch data from the data aggregators and respond to them by 
+scaling the application.  Auto-scalers are created and defined like other resources via REST endpoints and belong to the 
+namespace just as a `ReplicationController` or `Service`.
+
+Since an auto-scaler is a durable object it is best represented as a resource.  
+
+```go
+    //The auto scaler interface
+    type AutoScalerInterface interface {        
+        //ScaleApplication adjusts a resource's replica count.  Calls resize endpoint.  Args to this are based on what the endpoint
+        //can support.  See https://github.com/GoogleCloudPlatform/kubernetes/issues/1629
+        ScaleApplication(num int) error
+    }
+
+    type AutoScaler struct {
+        //common construct
+        TypeMeta
+        //common construct
+        ObjectMeta
+        
+        //Spec defines the configuration options that drive the behavior for this auto-scaler 
+        Spec    AutoScalerSpec  
+        
+        //Status defines the current status of this auto-scaler.
+        Status  AutoScalerStatus               
+     }
+     
+    type AutoScalerSpec struct {
+        //AutoScaleThresholds holds a collection of AutoScaleThresholds that drive the auto scaler
+        AutoScaleThresholds []AutoScaleThreshold
+        
+        //Enabled turns auto scaling on or off
+        Enabled boolean 
+        
+        //MaxAutoScaleCount defines the max replicas that the auto scaler can use.  This value must be greater than 
+        //0 and >= MinAutoScaleCount
```

Comment:
Is there any reason to not allow 0 as the max? Maybe someone would want it for testing or something.